fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/provider/foundationdb/OnlineIndexer.java

```diff
@@ -1338,22 +1342,29 @@ public boolean shouldAllowUniquePendingState(FDBRecordStore store) {
          */
         public boolean shouldAllowTypeConversionContinue(IndexBuildProto.IndexBuildIndexingStamp.Method newMethod,
                                                          IndexBuildProto.IndexBuildIndexingStamp.Method oldMethod) {
-            if (allowedTakeoverSet != null) {
-                if (newMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
-                    if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MULTI_TARGET_BY_RECORDS) {
-                        return allowedTakeoverSet.contains(TakeoverTypes.MULTI_TARGET_TO_SINGLE);
-                    }
-                    if (oldMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.MUTUAL_BY_RECORDS) {
-                        return allowedTakeoverSet.contains(TakeoverTypes.MUTUAL_TO_SINGLE);
-                    }
+            if (newMethod == IndexBuildProto.IndexBuildIndexingStamp.Method.BY_RECORDS) {
```

Comment:
So the total set of methods is:
```
  enum Method {
    INVALID      = 0;  // let zero be invalid
    BY_RECORDS   = 1;  // indexing by records - scanning all records in a record store
    BY_INDEX     = 2;  // indexing by a source index. When source index points to all possible records of the target index.
    SCRUB_REPAIR = 3;  // while readable, verify index validity. Repair if needed.
    MULTI_TARGET_BY_RECORDS = 4; // scan records, build multiple target indexes
    MUTUAL_BY_RECORDS = 5; // scan records, build multiple target indexes, while allowing multiple indexer processes
    NONE = 6; // return this stamp in query when null - never to be written in the DB
  };

```

And it would never make sense to go to, or from `INVALID` or `SCRUB_REPAIR`.
`BY_INDEX` to any of the others I think is not possible because the ranges would be wrong.
`NONE` to anything is implicitly supported (to the extent that `NONE` is a real type)

And the only other one that is not covered after this change `BY_RECORDS` to `MULTI_TARGET_BY_RECORDS`, although you could cause that to happen by first switching to `MUTUAL_BY_RECORDS`.

Reply:
As you've said, switching from `BY_INDEX` to any by-records method should never be allowed. 
Switching from `BY_RECORDS` to `MULTI_TARGET_BY_RECORDS` could have been allowed only if no records were scanned. 
It can make sense to verify that when switching from `BY_RECORDS` or `MULTI_TARGET_BY_RECORDS` to `MUTUAL_BY_RECORDS`, the very same target index set is being used. I'll try to add it.

Reply:
✅